src/include/RADOS/RADOS.hpp

```diff
@@ -0,0 +1,858 @@
+  auto enumerate_objects(std::int64_t pool, const EnumerationCursor& begin,
+			 const EnumerationCursor& end, const std::uint32_t max,
+			 const bufferlist& filter,
+			 std::vector<EnumeratedObject>* ls,
+			 EnumerationCursor* cursor,
+			 CompletionToken&& token,
+			 std::optional<std::string_view> ns = {},
+			 std::optional<std::string_view> key = {}) {
+    boost::asio::async_completion<CompletionToken, SimpleOpSig> init(token);
+    enumerate_objects(pool, begin, end, max, filter,
+		      ls, cursor,
+		      SimpleOpComp::create(get_executor(),
+					   std::move(init.completion_handler)),
+		      ns, key);
+    return init.result.get();
+  }
+
+  using CommandSig = void(boost::system::error_code,
+			  std::string, ceph::bufferlist);
+  using CommandComp = ceph::async::Completion<CommandSig>;
+  template<typename CompletionToken>
+  auto osd_command(int osd, std::vector<std::string>&& cmd,
+		   ceph::bufferlist&& in, CompletionToken&& token) {
+    boost::asio::async_completion<CompletionToken, CommandSig> init(token);
+    osd_command(osd, std::move(cmd), std::move(in),
+		CommandComp::create(get_executor(),
+				      std::move(init.completion_handler)));
+    return init.result.get();
+  }
+  template<typename CompletionToken>
+  auto pg_command(pg_t pg, std::vector<std::string>&& cmd,
+		  ceph::bufferlist&& in, CompletionToken&& token) {
+    boost::asio::async_completion<CompletionToken, CommandSig> init(token);
+    pg_command(pg, std::move(cmd), std::move(in),
+	       CommandComp::create(get_executor(),
+				      std::move(init.completion_handler)));
+    return init.result.get();
+  }
+  template<typename CompletionToken>
+  auto enable_application(std::string_view pool, std::string_view app_name,
+			  bool force, CompletionToken&& token) {
+    boost::asio::async_completion<CompletionToken, SimpleOpSig> init(token);
+    enable_application(pool, app_name, force,
+		       SimpleOpComp::create(get_executor(),
+					    std::move(init.completion_handler)));
+    return init.result.get();
+  }
+  uint64_t instance_id() const;
+
+private:
+
+  void execute(const Object& o, const IOContext& ioc, ReadOp&& op,
+	       ceph::buffer::list* bl, std::unique_ptr<Op::Completion> c,
+	       version_t* objver);
+
+  void execute(const Object& o, const IOContext& ioc, WriteOp&& op,
+	       std::unique_ptr<Op::Completion> c, version_t* objver);
+
+  void execute(const Object& o, std::int64_t pool, ReadOp&& op,
+	       ceph::buffer::list* bl, std::unique_ptr<Op::Completion> c,
+	       std::optional<std::string_view> ns,
+	       std::optional<std::string_view> key,
+	       version_t* objver);
+
+  void execute(const Object& o, std::int64_t pool, WriteOp&& op,
+	       std::unique_ptr<Op::Completion> c,
+	       std::optional<std::string_view> ns,
+	       std::optional<std::string_view> key,
+	       version_t* objver);
+
+  void lookup_pool(std::string name, std::unique_ptr<LookupPoolComp> c);
```

Comment:
Surely can. I did it originally since we hold a copy of the string in the case that we have to wait for the OSD map, but the caller doesn't really need to know that, and it's likely a fairly uncommon case anyway.